### PR TITLE
[REVIEW] Upgrade `pandas` version in `cudf`

### PR DESCRIPTION
**Pull request by galipremsagar**
Fixes: #9125 

This PR:

- [x] Upgrades pandas to its latest version - `1.3.2`
- [x] Fixes multiple hidden issues in `cudf` that were fixed as part of the latest `pandas` release.
- [x] Introduces `xfail` at multiple places which are essentially regressions in the latest `pandas` release.
- [x] Split up and refactor some `pytests` that will help us determine failures easily and are maintainable.

### CODE REVIEW

**Comment by galipremsagar** [comment]
rerun tests

File: ci/gpu/build.sh [hunk 1]
```diff
@@ -83,8 +83,8 @@ gpuci_mamba_retry install -y \
                   "ucx-py=0.22.*"
 
 # https://docs.rapids.ai/maintainers/depmgmt/
-# gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
-# gpuci_mamba_retry install -y "your-pkg=1.0.0"
+gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
+gpuci_mamba_retry install -y "pandas=1.3.2 arrow-cpp=5.0.0 arrow-cpp-proc * cuda pyarrow=5.0.0=*cuda"
```
Comment (galipremsagar) [inline ci/gpu/build.sh:87]:
```suggestion
gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda
```

File: ci/gpu/build.sh [hunk 2]
```diff
@@ -84,7 +84,7 @@ gpuci_mamba_retry install -y \
 
 # https://docs.rapids.ai/maintainers/depmgmt/
 # gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
```
Comment (galipremsagar) [inline ci/gpu/build.sh:86]:
```suggestion
gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
```

File: ci/gpu/build.sh [hunk 3]
```diff
@@ -83,8 +83,8 @@ gpuci_mamba_retry install -y \
                   "ucx-py=0.22.*"
 
 # https://docs.rapids.ai/maintainers/depmgmt/
-# gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
-# gpuci_mamba_retry install -y "your-pkg=1.0.0"
+gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
+gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda gmock
```
Comment (galipremsagar) [inline ci/gpu/build.sh:87]:
```suggestion
gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda gmock cython=0.29.24
```

File: ci/gpu/build.sh [hunk 4]
```diff
@@ -83,8 +83,8 @@ gpuci_mamba_retry install -y \
                   "ucx-py=0.22.*"
 
 # https://docs.rapids.ai/maintainers/depmgmt/
-# gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
-# gpuci_mamba_retry install -y "your-pkg=1.0.0"
+gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
+gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda gmock cython=0.29.24
```
Comment (galipremsagar) [inline ci/gpu/build.sh:87]:
```suggestion
gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda gmock cython=0.29.24 pytest pytest-benchmark pytest-xdist
```

File: ci/gpu/build.sh [hunk 5]
```diff
@@ -83,8 +83,8 @@ gpuci_mamba_retry install -y \
                   "ucx-py=0.22.*"
 
 # https://docs.rapids.ai/maintainers/depmgmt/
-# gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
-# gpuci_mamba_retry install -y "your-pkg=1.0.0"
+gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
+gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda gmock cython=0.29.24 pytest pytest-benchmark pytest-xdist
```
Comment (galipremsagar) [inline ci/gpu/build.sh:87]:
```suggestion
gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda gmock cython=0.29.24 pytest pytest-benchmark pytest-xdist pytest-cov pytest-timeout
```

File: ci/gpu/build.sh [hunk 6]
```diff
@@ -83,8 +83,8 @@ gpuci_mamba_retry install -y \
                   "ucx-py=0.22.*"
 
 # https://docs.rapids.ai/maintainers/depmgmt/
-# gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
-# gpuci_mamba_retry install -y "your-pkg=1.0.0"
+gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
+gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda gmock cython=0.29.24 pytest pytest-benchmark pytest-xdist pytest-cov pytest-timeout cupy==9.4.0 cachetools transformers nvtx=0.2.3 protobuf packaging mimesis=4.0.0 hypothesis rapidjson double-conversion dlpack=0.5 streamz typing_extensions mypy=0.782 pandoc=1.19.2 fsspec=2021.8.1 fastavro=1.4.4 numba=0.53.1 pyorc python-confluent-kafka=1.6.0
```
Comment (galipremsagar) [inline ci/gpu/build.sh:87]: A temporary change which will be reverted once @rapidsai/ops-codeowners merges https://github.com/rapidsai/integration/pull/345

Comment (galipremsagar) [reply ci/gpu/build.sh]: Reverted.

**Review by Ethyling — APPROVED** [review]

File: python/cudf/cudf/tests/test_dataframe.py [hunk 7]
```diff
@@ -1259,7 +1241,8 @@ def test_dataframe_concat_different_numerical_columns(dtype1, dtype2):
     else:
         pres = pd.concat([df1, df2])
         gres = cudf.concat([cudf.from_pandas(df1), cudf.from_pandas(df2)])
-        assert_eq(cudf.from_pandas(pres), gres)
+        # Pandas 1.3.2+ returns mixed `object` dtype result.
+        assert_eq(cudf.from_pandas(pres.astype(gres.dtypes)), gres)
```
Comment (shwina) [inline python/cudf/cudf/tests/test_dataframe.py:1245]: I think in situations like this, it's cleaner to use `assert_eq(expect, got, check_dtypes=False)` -- or does that not achieve the same thing?

File: python/cudf/cudf/tests/test_dataframe.py [hunk 8]
```diff
@@ -1764,15 +1747,20 @@ def test_dataframe_transpose(nulls, num_cols, num_rows, dtype):
     pdf = pd.DataFrame()
 
     null_rep = np.nan if dtype in ["float32", "float64"] else None
-
```
Comment (shwina) [inline python/cudf/cudf/tests/test_dataframe.py:1767]: Not really understanding the changes in this test -- can we add a comment to explain what's going on?

File: python/cudf/cudf/tests/test_index.py [hunk 9]
```diff
@@ -1866,7 +1873,12 @@ def test_index_fillna(data, fill_value):
     pdi = pd.Index(data)
     gdi = cudf.Index(data)
 
-    assert_eq(pdi.fillna(fill_value), gdi.fillna(fill_value))
```
Comment (shwina) [inline python/cudf/cudf/tests/test_index.py:1869]: We could just use `check_dtypes=False` here too.

**Review by shwina — COMMENTED** [review]
This is invaluable work, @galipremsagar!

File: python/cudf/cudf/tests/test_dataframe.py (hunk 8 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py:1767]: Sure, so this was a change in behavior incase of `bool` column:

Until pandas `1.2.5`, setting a `nan` to bool column would result in conversion to `float`:
```python
s = pd.Series([True, False, True])
s
0     True
1    False
2     True
dtype: bool
s[[2]] = np.nan
s
0    1.0
1    0.0
2    NaN
dtype: float64
```

Now that's fixed to conversion to an `object` dtype:

```python
>>> s = pd.Series([True, False, True])
>>> s
0     True
1    False
2     True
dtype: bool
>>> s[[2]] = np.nan
>>> s
0     True
1    False
2      NaN
dtype: object
```

So in `cudf` we change `object` dtype to `str` type - for which there is no `transpose` implemented yet.

Hence the fix I have made in this pytest is to correctly cast the data to pandas nullable types and perform a transpose

Comment (shwina) [reply python/cudf/cudf/tests/test_dataframe.py:1767]: Ahh, thanks! Let's add a comment in the code here for our future selves?

File: python/cudf/cudf/tests/test_index.py (hunk 9 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_index.py:1869]: `assert_index_equal` doesn't seem to be having that parameter. The closest thing that skips `type` checking is `exact` param: https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_index_equal.html

File: python/cudf/cudf/tests/test_dataframe.py (hunk 7 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py]: Yup, changed this.

File: python/cudf/cudf/tests/test_dataframe.py (hunk 8 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py:1767]: Done 👍

**Review by shwina — APPROVED** [review]

**Comment by galipremsagar** [comment]
@shwina I made a trivial fix to a failure being caused by latest `dask` upstream changes: https://github.com/rapidsai/cudf/pull/9147/commits/1ded3a3ca99bd59d38b749f043a16e8879b08c9a

This was to match pandas `get_slice_bound` function signature.

**Review by shwina — APPROVED** [review]

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**